lua/entities/obj_vj_npccontroller/init.lua

```diff
@@ -257,189 +262,210 @@ net.Receive("vj_controller_cldata", function(len, ply)
+		end
+		net.WriteInt(bone, 10)
+		net.WriteBool((reset != true and npcData.FirstP_ShrinkBone) or false)
+		net.WriteUInt((reset != true and npcData.FirstP_CameraBoneAng) or 0, 2)
+		net.WriteInt((reset != true and npcData.FirstP_CameraBoneAng_Offset) or 0, 10)
+	net.Send(ply)
 end
 ---------------------------------------------------------------------------------------------------------------------------------------------
 local vecZ20 = Vector(0, 0, 20)
+local curTime = CurTime()
+local defAttackTypes = {MeleeAttack=false, RangeAttack=false, LeapAttack=false, WeaponAttack=false, GrenadeAttack=false, Ammo="---"}
 --
 function ENT:Think()
-	if (!self.VJCE_Camera:IsValid()) then self:StopControlling() return end
-	if !IsValid(self.VJCE_Player) /*or self.VJCE_Player:KeyDown(IN_USE)*/ or self.VJCE_Player:Health() <= 0 or (!self.VJCE_Player.IsControlingNPC) or !IsValid(self.VJCE_NPC) or (self.VJCE_NPC:Health() <= 0) then self:StopControlling() return end
-	if self.VJCE_Player.IsControlingNPC != true then return end
-	if self.VJCE_Player.IsControlingNPC && IsValid(self.VJCE_NPC) then
-		self.VJC_NPC_LastPos = self.VJCE_NPC:GetPos()
-		self.VJCE_Player:SetPos(self.VJC_NPC_LastPos + vecZ20) -- Set the player's location
+	local ply = self.VJCE_Player
+	local npc = self.VJCE_NPC
+	local camera = self.VJCE_Camera
+	if (!camera:IsValid()) then self:StopControlling() return end
+	if !IsValid(ply) /*or ply:KeyDown(IN_USE)*/ or ply:Health() <= 0 or (!ply.IsControlingNPC) or !IsValid(npc) or (npc:Health() <= 0) then self:StopControlling() return end
+	if ply.IsControlingNPC != true then return end
+	if ply.IsControlingNPC && IsValid(npc) then
+		local npcWeapon = npc:GetActiveWeapon()
+		self.VJC_NPC_LastPos = npc:GetPos()
+		ply:SetPos(self.VJC_NPC_LastPos + vecZ20) -- Set the player's location
 		self:SendDataToClient()
 		
 		-- HUD
-		local AttackTypes = {MeleeAttack=false, RangeAttack=false, LeapAttack=false, WeaponAttack=false, GrenadeAttack=false, Ammo="---"}
-		if self.VJCE_NPC.IsVJBaseSNPC == true then
-			if self.VJCE_NPC.HasMeleeAttack == true then AttackTypes["MeleeAttack"] = ((self.VJCE_NPC.IsAbleToMeleeAttack != true or self.VJCE_NPC.AttackType == VJ_ATTACK_MELEE) and 2) or true end
-			if self.VJCE_NPC.HasRangeAttack == true then AttackTypes["RangeAttack"] = ((self.VJCE_NPC.IsAbleToRangeAttack != true or self.VJCE_NPC.AttackType == VJ_ATTACK_RANGE) and 2) or true end
-			if self.VJCE_NPC.HasLeapAttack == true then AttackTypes["LeapAttack"] = ((self.VJCE_NPC.IsAbleToLeapAttack != true or self.VJCE_NPC.AttackType == VJ_ATTACK_LEAP) and 2) or true end
-			if IsValid(self.VJCE_NPC:GetActiveWeapon()) then AttackTypes["WeaponAttack"] = true AttackTypes["Ammo"] = self.VJCE_NPC:GetActiveWeapon():Clip1() end
-			if self.VJCE_NPC.HasGrenadeAttack == true then AttackTypes["GrenadeAttack"] = (CurTime() <= self.VJCE_NPC.NextThrowGrenadeT and 2) or true end
+		local AttackTypes = defAttackTypes -- Optimization?
```

Comment:
If it's not redeclared, the table will not update accordingly, ex: NPC has a weapon, so when attack becomes true and ammo counter visible, when the NPC's weapon is removed, the controller will continue to think it has a weapon because the table value remains set to true